Feature: Easymerchant payment gateway

manoj-easymerchant opened the pull request:
Hello,
This is the first PR for invoiceninja.

API Documentation : https://developers.lyfepay.io
Website: https://lyfecyclex.com

I have added the Easymerchant gateway that will support both Card and ACH bank transfer payments.

Kindly review the PR and let me know the suggestion or feedback.  

Note:
I have updated the mentioned feedback points . And also added Refund functionality, ACH webhook status change added

manoj-easymerchant committed:
added easymerchant gateway

manoj-easymerchant committed:
removed blank space

manoj-easymerchant committed:
test comment

manoj-easymerchant committed:
test comment

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/invoiceninja/invoiceninja?pullRequest=9045) <br/>All committers have signed the CLA.

turbo124 commented on app/Models/PaymentType.php:
```diff
@@ -146,6 +146,7 @@ public static function parseCardType($cardName)
             'maestro' => self::MAESTRO,
             'solo' => self::SOLO,
             'switch' => self::SWITCH,
+            'ach' => self::ACH,
```
not required, already exists in payment types.

turbo124 commented on database/migrations/2023_11_28_103705_easymerchant.php:
```diff
@@ -0,0 +1,43 @@
+            $gateway = new Gateway();
+            $gateway->id = 63;
+            $gateway->name = 'Easymerchant';
+            $gateway->key = Str::lower(Str::random(32));
```
this needs to be a static string, otherwise it will be different on every installation.

manoj-easymerchant replied:
okay sure. let me make it to static data

turbo124 commented on app/PaymentDrivers/EasymerchantPaymentDriver.php:
```diff
@@ -0,0 +1,107 @@
+    public function gatewayTypes(): array
+    {
+        $types = [];
+
+        $types[] = GatewayType::CREDIT_CARD;
+        $types[] = GatewayType::BANK_TRANSFER;
+
```
If the gateway only supports USD, you'll want to make sure the clients currency is USD to prevent the wrong gateways / inappropriate gateways being displayed.

manoj-easymerchant replied:
yep got it. we added to it

turbo124 reviewed (commented):
Thank for submitting this PR, is there an actual webpage that one can view for this Gateway? If you can provide a URL that we can use to view this Payment Processor please.

manoj-easymerchant committed:
resolved feedback points. refund & webhook functionality is added

manoj-easymerchant commented:
Dear @turbo124 ,
Thanks for the quick response and the feedbacks.

Here i have updated the mentioned points and also added the refund functionality, webhook for ACH status update has been implemented.
Kindly review and let me know the improvement points and changes .

API Documentation : https://developers.lyfepay.io/
Thank you.

manoj-easymerchant committed:
Merge branch 'v5-develop' into Easymerchant

Signed-off-by: manoj-easymerchant <114509453+manoj-easymerchant@users.noreply.github.com>

turbo124 commented:
Hi there,

Aftering reviewing the website and this PR, we won't be able to merge this PR. 

The implementation here is passing naked credit card numbers _through_ Invoice Ninja. This would require PCI compliance that allows the storage/transmission of Credit Card Numbers which is very expensive from both a financial and compliance aspect.

We would only consider integrations that use a tokenization process, where all credit card details are communicated over the processors network and only a payment token would be handled by Invoice Ninja.
